### ISSUE TITLE
Set PIR fails with `Service Unavailable` for Argus Eco

### ISSUE THREAD

**Issue by fmarzocca** [QuantumEntangledAndy/neolink#51]
My camera is an Argus Eco. I am trying to switch OFF the PIR:

```
./neolink pir --config=neolink_config.toml giardino off
[2023-03-31T07:56:20Z INFO  neolink] Neolink 4fcaf4d538dc6e05434690bdf7f3a3bc22b077e3 release
[2023-03-31T07:56:20Z INFO  neolink::utils] giardino: Connecting to camera at UID: 95270004YFGE16SE
[2023-03-31T07:56:22Z INFO  neolink_core::bc_protocol] Relay success 95270004YFGE16SE at 35.180.129.86:51736
[2023-03-31T07:56:22Z INFO  neolink::utils] giardino: Logging in
[2023-03-31T07:56:24Z INFO  neolink::utils] giardino: Connected and logged in
Error: Unable to set camera PIR state

Caused by:
    Camera responded with Service Unavaliable
```

**Comment by QuantumEntangledAndy** [comment]
Enable trace level logging and try again.

```bash
export RUST_LOG=trace
```

Then post the log here so I can see what the camera sent in reply.

**Comment by fmarzocca** [comment]
[Here](https://gist.github.com/fmarzocca/6388d99a19624192a2bf36d6978f629f) it is.

**Comment by QuantumEntangledAndy** [comment]
As far as I can tell the camera is replying that it lacks the capability to set what we requested. I think perhaps we have to send a different set of commands for this model. I'll try capturing the command with my eco2e maybe it's the same as the eco.

**Comment by QuantumEntangledAndy** [comment]
Just to confirm this is what we want to turn on/off

<img width="520" alt="Screenshot 2023-03-31 at 15 31 00" src="https://user-images.githubusercontent.com/13386481/229068611-a81fa51b-7023-42b1-b68f-9facd2cc2835.png">

**Comment by fmarzocca** [comment]
> Just to confirm this is what we want to turn on/off

I have a different image:
![alt text](https://www.socrates-server.com/neolink_test.png)

**Comment by fmarzocca** [comment]
> This is the message for controlling this:

Let me know if I can make some test for you, to query the commands list

**Comment by QuantumEntangledAndy** [comment]
Best thing would be wireshark. Do you know how to use wireshark?

If so have wireshark running on your computer capturing the network traffic then open the official client (on the same computer as wireshark) and disable/enable the pir using that. It will record the packets sent. Filter the packets to only save those going to/from the camera then send the dump to me as a pcap.

**Comment by fmarzocca** [comment]
Attached.
[ArgusEco.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11123662/ArgusEco.pcap.zip)

**Comment by QuantumEntangledAndy** [comment]
Thanks but unfortunalty your your wireshark dump only contain 18 packets and dosen't have anything to do with either login or changing the PIR. Maybe you filtered out too much or started the capture after already  opening the offical client.

**Comment by fmarzocca** [comment]
Weird. I commonly use wireshark. I have filtered for packets going to/from the camera IP. And I started recording before I switched the PIR.

i can try another capture.

**Comment by QuantumEntangledAndy** [comment]
Perhaps you can install our wireshark dissector. It will let you filter on baichuan protocol packets

**Comment by QuantumEntangledAndy** [comment]
This is my current filter
```
baichuan.msg_id && baichuan.msg_id != 3 && baichuan.msg_id != 234 && baichuan.msg_id != 115 && baichuan.msg_id != 33
```

it will remove al lot of the data we are not interested in like camera feeds and motion detection and wifistatus updates etc

**Comment by QuantumEntangledAndy** [comment]
This is the message for reqesting the status this:

- Client

|    Magic         |  Message ID  | Message Length | Encryption Offset |    Status Code    | Message Class | Payload Offset |
|--------------|--------------|------------------|-------------------|-------------------|---------------|---------------|
| f0 de bc 0a  | d4 00 00 00   | 5e 00 00 00         | 00 00 00 13         | 00 00                    | 14 64              | 5e 00 00 00 |

**Meta Payload:**

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<rfId>0</rfId>
</Extension>
```

- Camera

|    Magic         |  Message ID  | Message Length | Encryption Offset |    Status Code    | Message Class | Payload Offset |
|--------------|--------------|------------------|-------------------|-------------------|---------------|---------------|
| f0 de bc 0a  | d4 00 00 00   | a6 04 00 00        | 00 00 00 13         | c8 00                    | 00 00             | 00000000 |

**Main Payload:**

```xml
<?xml version="1.0" encoding="UTF-8" ?>
    <body>
    <rfAlarmCfg version="1.1">
    <rfID>0</rfID>
    <enable>1</enable>
    <sensitivity>0</sensitivity>
    <sensiValue>21</sensiValue>
    <reduceFalseAlarm>0</reduceFalseAlarm>
    <timeBlockList>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Sunday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Monday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Tuesday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Wednesday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Thursday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Friday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    <timeBlock>
    <enable>1</enable>
    <weekDay>Saturday</weekDay>
    <beginHour>0</beginHour>
    <endHour>23</endHour>
    </timeBlock>
    </timeBlockList>
    <alarmHandle>
    <item>
    <channel>0</channel>
    <handleType>snap,rec,push</handleType>
    </item>
    </alarmHandle>
    </rfAlarmCfg>
    </body>
```

**Comment by fmarzocca** [comment]
A better wireshark capture.... attached
[ArgusEco2.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11129044/ArgusEco2.pcap.zip)

**Comment by fmarzocca** [comment]
> 

don't know how

**Comment by fmarzocca** [comment]
> Perhaps you can install our wireshark dissector. It will let you filter on baichuan protocol packets

don't know how

**Comment by QuantumEntangledAndy** [comment]
This is the file https://github.com/QuantumEntangledAndy/neolink/blob/master/dissector/baichuan.lua

It need to be put in the right folder. In mac/linux it is `${HOME}/.local/lib/wireshark/plugins/baichuan.lua`

On windows it is `%APPDATA%\Wireshark\plugins`

**Comment by QuantumEntangledAndy** [comment]
Looking at your dump there are a few interesting packets 299 and 342. They seem to be sent right after you request the PIR status and before the setting of the PIR. However I cannot decrypt there contents without a password. Perhaps you could set a dummy password during the dump of `123456` and then capture it again. I will need the login message `1` to get the nonce too since the AES encryption key it `PASSWORD-NONCE`

**Comment by fmarzocca** [comment]
You can use 140289 to decrypt. I have changed the password.

**Comment by QuantumEntangledAndy** [comment]
# 299

- Client

f0debc0a2b010000680000002f0000000000146468000000

**Meta Payload:**

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<channelId>0</channelId>
</Extension>
```

- Camera:
No reply

# 342

- Client

f0debc0a56010000eb000000300000000000146468000000

**Meta Payload:**

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<channelId>0</channelId>
</Extension>
```

**Main Payload:**

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<body>
<AiDetectCfg version="1.1">
<chn>0</chn>
<type>people</type>
</AiDetectCfg>
</body>
```

**Notes:**
Theres more than one possible payload. This has also been observed:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<body>
<AiDetectCfg version="1.1">
<chn>0</chn>
<type>vehicle</type>
</AiDetectCfg>
</body>
```

- Camera

f0debc0a560100006e03000030000000c800000000000000

**Main Payload:**

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<body>
<AiDetectCfg version="1.1">
<chn>0</chn>
<type>people</type>
<sensitivity>69</sensitivity>
<stayTime>0</stayTime>
<minTargetHeight>0.30354199</minTargetHeight>
<minTargetWidth>0.17060199</minTargetWidth>
<maxTargetHeight>0</maxTargetHeight>
<maxTargetWidth>0</maxTargetWidth>
<width>72</width>
<height>40</height>
<area>AAAD8AB4AAxgAAAD8AAAAABgAAAD8AAAAAABAAAD//gAAAADAEAD//gAAAAAAAAf//4AAAAAAAAf//8AAAAAAAAf//+AOAABAAAf///AAAABAAD////4AAABAAH////+AAABAAP/////wBAAAAf/////+AAAAA///////gAAAA///////4AAkB///////+AA//////////gA//////////4A//////////8D//////////8B//////////+B///////////x////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////</area>
</AiDetectCfg>
</body>
```

---

**Other differences:**

# 212 (AlarmCfg)

You had no reply for your AlarmCfg message

**Comment by QuantumEntangledAndy** [comment]
You didn't seem to get a reply about the PIR detector in the same way I did. But you did get one about AI detection. Do you know if your camera has PIR or does it detect just by AI on vid feed

**Comment by fmarzocca** [comment]
It has a large PIR on the front:

[image](https://manuals.plus/wp-content/uploads/2021/08/Reolink-Argus-Eco-featured-image.jpg)

**Comment by fmarzocca** [comment]
What are msgs 192 and 319?

**Comment by fmarzocca** [comment]
> . But you did get one about AI detection. 

Vehicle detection is disabled, so it should not be correctly decoded.

**Comment by QuantumEntangledAndy** [comment]
Vehicle detection decodes ok. It's just marked as disabled in the return message from the camera.

Just to confirm you did turn the PIR on and off during your tests right.

I'll check the other messages maybe I missed something.

**Comment by fmarzocca** [comment]
Yes, I did.


- Sent from my mobile phone. Sorry for typos -

Il Sab 1 Apr 2023, 16:05 Andrew King ***@***.***> ha scritto:

**Comment by fmarzocca** [comment]
Could you pls check if the attached 2 messages could be PIR ON and PIR OFF?
[Argus_Eco_3.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11132039/Argus_Eco_3.pcap.zip)

**Comment by QuantumEntangledAndy** [comment]
Yes 213 is pir on off. It's the message that we already send. When we send the use the neolink pir on/off command. 382 was if I recall header only with no data attached, so it's hard to know what it is for.

**Comment by fmarzocca** [comment]
But if 213 is PIR, why I found it in the stream but it doesn't work if I send from noelink?

**Comment by QuantumEntangledAndy** [comment]
That's what we are trying to figure out. I have a test build I'm working on that more closely follows the format of Reolink. Will hopefully send it to you later

**Comment by QuantumEntangledAndy** [comment]
If this fails to work please capture the packets sent by neolink and we will compare them to the offical

**Comment by fmarzocca** [comment]
> OK the PR to address this is here #53. Once it comiples you can get the builds from here https://github.com/QuantumEntangledAndy/neolink/actions/runs/4594196290

Thanks! How much does it normally take to build?

**Comment by QuantumEntangledAndy** [comment]
Looks like it is already built. The artefacts are at the bottom on that link (make sure your logged in to GitHub)

**Comment by fmarzocca** [comment]
Doesn't work either. How can I grab Wireshark streams? neolink is on a Raspberry in the LAN, not on the computer from which Wireshark is initiated.

**Comment by fmarzocca** [comment]
I figured it out, with tcpdump. But I got only few packets. Attached.
[tcp.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11136682/tcp.pcap.zip)

**Comment by fmarzocca** [comment]
I don't see any 213 msg in the stream...

**Comment by fmarzocca** [comment]
This is another stream with 213 msg! I verified that randomly the command works, but not every time. Some time it fails, and some other it works.
[tcp.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11136833/tcp.pcap.zip)

**Comment by QuantumEntangledAndy** [comment]
Does it always fail with CameraServiceUnavaliable or is there a new message. Now the first pcap you sent seems to be successful. I'll check the second one now.

p.s. You can now also use
```bash
neolink pir --config=config.toml
```
witout on/off to see the current status without setting

**Comment by QuantumEntangledAndy** [comment]
Sorry other way around your second pcap is successful.

Your first seems to fail to get the current PIR status

**Comment by QuantumEntangledAndy** [comment]
With status code 400

**Comment by fmarzocca** [comment]
After 2 or 3 times:

```
./neolink pir giardino --config=neolink_config.toml
[2023-04-03T10:04:47Z INFO  neolink] Neolink 99c624137263503aa3f2fc254d89c5189947a0b6 release
[2023-04-03T10:04:47Z INFO  neolink::utils] giardino: Connecting to camera at UID: 95270004YFGE16SE
[2023-04-03T10:04:47Z INFO  neolink_core::bc_protocol] Local discovery success 95270004YFGE16SE at 192.168.1.42:22034
[2023-04-03T10:04:47Z INFO  neolink::utils] giardino: Logging in
[2023-04-03T10:04:48Z INFO  neolink::utils] giardino: Connected and logged in
<?xml version="1.0" encoding="utf-8"?><RfAlarmCfg version="1.1"><rfID>0</rfID><enable>1</enable><sensitivity>0</sensitivity><sensiValue>12</sensiValue><reduceFalseAlarm>0</reduceFalseAlarm><timeBlockList /><alarmHandle /></RfAlarmCfg>
```

**Comment by QuantumEntangledAndy** [comment]
I have an idea. Please wait for next  build

**Comment by fmarzocca** [comment]
Here attached. Pwd: 123456
[tcp.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11137110/tcp.pcap.zip)

**Comment by fmarzocca** [comment]
Password 123456
[tcp.pcap.zip](https://github.com/QuantumEntangledAndy/neolink/files/11137147/tcp.pcap.zip)

**Comment by QuantumEntangledAndy** [comment]
I noticed in our extension xml neolink does this

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<channelId>0</channelId>
</Extension>
```

while reolink does this
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<rfId>0</rfId>
</Extension>
```

Just swapped it to rfID in the next build here https://github.com/QuantumEntangledAndy/neolink/actions/runs/4595507575

**Comment by QuantumEntangledAndy** [comment]
Not sure if there is a better way to get this rdId field then just setting it to the channelId. Maybe need to look for that

**Comment by fmarzocca** [comment]
This is a GIF showing the connection time when I launch the Reolink Client
![ArgusEco](https://user-images.githubusercontent.com/8530030/229499054-491ce79e-e134-424e-883d-1c9bcd3fd3cf.gif)

**Comment by QuantumEntangledAndy** [comment]
Maybe this has something to do with the battery camera waking up? The camera usually sends a bunch of messages such as the capabilities after login so maybe I should wait for some of those

**Comment by QuantumEntangledAndy** [comment]
I notice that 212 fails if it sends before these capability messages

**Comment by fmarzocca** [comment]
I think you are right

**Comment by fmarzocca** [comment]
Maybe you should wait for the capability messages before sending the command.

**Comment by QuantumEntangledAndy** [comment]
I'm working on creating something to properly parse tha capability. Since we can use it to give useful error messages such as `this camera dosent support pir`

**Comment by fmarzocca** [comment]
After many tests, I can confirm this: your neolink is correctly working only if I keep the connection alive (either through Client on computer or App on smartphone).

**Comment by fmarzocca** [comment]
the build failed

**Comment by QuantumEntangledAndy** [comment]
Ah sorry forgot to commit the new file and not just the modified ones.

**Comment by QuantumEntangledAndy** [comment]
Ok on its way https://github.com/QuantumEntangledAndy/neolink/actions/runs/4607363019

**Comment by fmarzocca** [comment]
It works!
Great job!

**Comment by QuantumEntangledAndy** [comment]
Are we all good? If so I'll close  this and merge

**Comment by fmarzocca** [comment]
I need to send the command twice in order to get it working everytime. It seems that first command needs to set the camera alive (failing) and the second goes through.
Sometime one command is enough, but mostly not.

**Comment by QuantumEntangledAndy** [comment]
Ok then, since the error code is always 400. I will see about retrying the messages if we get that error code.

**Comment by QuantumEntangledAndy** [comment]
Ok retry build is here https://github.com/QuantumEntangledAndy/neolink/actions/runs/4615033288

**Comment by QuantumEntangledAndy** [comment]
Reties every 0.5s for 2.5s

**Comment by fmarzocca** [comment]
Perfect! This is very good, no failures.
Yes, you can merge this.

**Comment by QuantumEntangledAndy** [comment]
Merged in #53
